Tribler/community/multichain/block.py

```diff
@@ -0,0 +1,282 @@
+            self.signature = EMPTY_SIG
+            # debug stuff
+            self.insert_time = None
+        else:
+            (self.up, self.down, self.total_up, self.total_down, self.public_key, self.sequence_number,
+             self.link_public_key, self.link_sequence_number, self.previous_hash, self.signature,
+             self.insert_time) = (data[0], data[1], data[2], data[3], data[4], data[5], data[6], data[7], data[8],
+                                  data[9], data[10])
+            if isinstance(self.public_key, buffer):
+                self.public_key = str(self.public_key)
+            if isinstance(self.link_public_key, buffer):
+                self.link_public_key = str(self.link_public_key)
+            if isinstance(self.previous_hash, buffer):
+                self.previous_hash = str(self.previous_hash)
+            if isinstance(self.signature, buffer):
+                self.signature = str(self.signature)
+
+    @property
+    def hash(self):
+        return sha256(self.pack()).digest()
+
+    def validate(self, database):
+        """
+        Validates this block against what is known in the database
+        :param database: the database to check against
+        :return: "valid" if the block does not violate any rules,
+                 "partial-next" if the block does not violate any rules, but there are missing blocks in the future
+                 "partial-prev" if the block does not violate any rules, but there are missing blocks in the past
+                 "partial" if the block does not violate any rules, but there are missing blocks on either side.
+                 "no-data" if there is not enough information known about the block to validate
+                 "invalid" if the block violate any of the rules,
+        """
+
+        result = ["valid"]
+        errors = []
+        crypto = ECCrypto()
+
+        def err(reason):
+            result[0] = "invalid"
+            errors.append(reason)
+
+        # Step 1: get all related blocks from the database, assume that the database speaks the truth and that all
+        # retrieved blocks are not invalid themselves. Thus the database should reject adding blocks that have "invalid"
+        # as validation result.
+        blk = database.get(self.public_key, self.sequence_number)
+        link = database.get_linked(self)
+        prev_blk = (database.get_blocks_until(self.public_key, self.sequence_number - 1, limit=1) or [None])[0]
+        next_blk = (database.get_blocks_since(self.public_key, self.sequence_number + 1, limit=1) or [None])[0]
+
+        # Step 2: determine the maximum validation level
+        if not prev_blk and not next_blk:
+            if self.sequence_number != GENESIS_SEQ and self.previous_hash != GENESIS_ID:
+                # No blocks found, there is no info to base on
+                err("No blocks are known for this member before or after the queried sequence number")
+                result[0] = "no-info"
```

Comment:
Hummm nice semantic point to consider here? Are we missing data or info? I choose no-info for now.